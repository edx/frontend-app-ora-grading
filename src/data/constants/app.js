import { getConfig } from '@edx/frontend-platform';

export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
export const locationId = window.location.pathname.slice(1);
