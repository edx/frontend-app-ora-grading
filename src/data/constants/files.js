import { StrictDict } from 'utils';

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
});

export default FileTypes;
