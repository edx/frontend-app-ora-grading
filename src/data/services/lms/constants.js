import { StrictDict } from 'utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const gradingStatusDisplay = StrictDict({
  [gradingStatuses.ungraded]: 'Ungraded',
  [gradingStatuses.locked]: 'Currently being graded by someone else',
  [gradingStatuses.graded]: 'Grading Complete',
  [gradingStatuses.inProgress]: 'You are currently grading this response',
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});
