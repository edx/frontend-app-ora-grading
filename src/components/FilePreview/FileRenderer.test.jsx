import React from 'react';
import { shallow } from 'enzyme';

import { FileTypes } from 'data/constants/files';
import {
  ImageRenderer,
  PDFRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';
import {
  FileRenderer, getFileType, ERROR_STATUSES, RENDERERS,
} from './FileRenderer';

jest.mock('./FileCard', () => 'FileCard');

jest.mock('components/FilePreview/BaseRenderers', () => ({
  PDFRenderer: () => 'PDFRenderer',
  ImageRenderer: () => 'ImageRenderer',
  TXTRenderer: () => 'TXTRenderer',
}));

jest.mock('./Banners', () => ({
  ErrorBanner: () => 'ErrorBanner',
  LoadingBanner: () => 'LoadingBanner',
}));

describe('FileRenderer', () => {
  describe('component', () => {
    const supportedTypes = [
      FileTypes.pdf,
      FileTypes.jpg,
      FileTypes.jpeg,
      FileTypes.bmp,
      FileTypes.png,
      FileTypes.txt,
    ];
    const files = [
      ...supportedTypes.map((fileType, index) => ({
        name: `fake_file_${index}.${fileType}`,
        description: `file description ${index}`,
        downloadUrl: `/url-path/fake_file_${index}.${fileType}`,
      })),
    ];

    const els = files.map((file) => shallow(<FileRenderer file={file} />));

    describe('snapshot', () => {
      els.forEach((el) => {
        const file = el.prop('file');
        const fileType = getFileType(file.name);

        test(`successful rendering ${fileType}`, () => {
          // I prefer to invoke this method for changing state.
          // It is consistence behavior even though it is not as pretty in snapshot.
          el.instance().onSuccess();
          expect(el.instance().render()).toMatchSnapshot();
        });
      });

      Object.keys(ERROR_STATUSES).forEach((status) => {
        test(`has error ${status}`, () => {
          const el = shallow(<FileRenderer file={files[0]} />);
          el.instance().onError(ERROR_STATUSES[status]);
          expect(el.instance().render()).toMatchSnapshot();
        });
      });
    });

    describe('component', () => {
      describe('uses the correct renderers', () => {
        const checkFile = (index, expectedRenderer) => {
          const file = files[index];
          const el = shallow(<FileRenderer file={file} />);
          const renderer = el.find(expectedRenderer);
          expect(renderer).toBeDefined();
          expect(renderer.prop('url')).toEqual(file.downloadUrl);
          expect(renderer.prop('fileName')).toEqual(file.name);
        };
        // the manual process for this is prefer.
        test(FileTypes.pdf, () => checkFile(0, PDFRenderer));
        test(FileTypes.jpg, () => checkFile(1, ImageRenderer));
        test(FileTypes.jpeg, () => checkFile(2, ImageRenderer));
        test(FileTypes.bmp, () => checkFile(3, ImageRenderer));
        test(FileTypes.png, () => checkFile(4, ImageRenderer));
        test(FileTypes.txt, () => checkFile(5, TXTRenderer));
      });
    });

    describe('renderer constraints', () => {
      els.forEach((el) => {
        const file = el.prop('file');
        const fileType = getFileType(file.name);
        const RendererComponent = RENDERERS[fileType];
        const ActualRendererComponent = jest.requireActual(
          'components/FilePreview/BaseRenderers',
        )[RendererComponent.name];

        it(`${fileType} renderer must have onError and onSuccess props`, () => {
          /* eslint-disable react/forbid-foreign-prop-types */
          expect(ActualRendererComponent.propTypes.onError).toBeDefined();
          expect(ActualRendererComponent.propTypes.onSuccess).toBeDefined();
        });
      });
    });
  });
});
