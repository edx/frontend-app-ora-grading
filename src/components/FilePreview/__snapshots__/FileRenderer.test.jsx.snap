// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileRenderer component snapshot has error 404 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 0",
      "downloadUrl": "/url-path/fake_file_0.pdf",
      "name": "fake_file_0.pdf",
    }
  }
>
  <ErrorBanner
    actions={
      Array [
        Object {
          "id": "retry",
          "message": Object {
            "defaultMessage": "Retry",
            "description": "Retry button for error in file renderer",
            "id": "ora-grading.ResponseDisplay.FileRenderer.retryButton",
          },
          "onClick": [Function],
        },
      ]
    }
    headingMessage={
      Object {
        "defaultMessage": "File not found",
        "description": "File not found error message",
        "id": "ora-grading.ResponseDisplay.FileRenderer.fileNotFound",
      }
    }
  >
    <FormattedMessage
      defaultMessage="File not found"
      description="File not found error message"
      id="ora-grading.ResponseDisplay.FileRenderer.fileNotFound"
    />
  </ErrorBanner>
</FileCard>
`;

exports[`FileRenderer component snapshot has error 500 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 0",
      "downloadUrl": "/url-path/fake_file_0.pdf",
      "name": "fake_file_0.pdf",
    }
  }
>
  <ErrorBanner
    actions={
      Array [
        Object {
          "id": "retry",
          "message": Object {
            "defaultMessage": "Retry",
            "description": "Retry button for error in file renderer",
            "id": "ora-grading.ResponseDisplay.FileRenderer.retryButton",
          },
          "onClick": [Function],
        },
      ]
    }
    headingMessage={
      Object {
        "defaultMessage": "Unknown errors",
        "description": "Unknown errors message",
        "id": "ora-grading.ResponseDisplay.FileRenderer.unknownError",
      }
    }
  >
    <FormattedMessage
      defaultMessage="Unknown errors"
      description="Unknown errors message"
      id="ora-grading.ResponseDisplay.FileRenderer.unknownError"
    />
  </ErrorBanner>
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering bmp 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 3",
      "downloadUrl": "/url-path/fake_file_3.bmp",
      "name": "fake_file_3.bmp",
    }
  }
>
  <ImageRenderer
    fileName="fake_file_3.bmp"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_3.bmp"
  />
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering jpeg 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 2",
      "downloadUrl": "/url-path/fake_file_2.jpeg",
      "name": "fake_file_2.jpeg",
    }
  }
>
  <ImageRenderer
    fileName="fake_file_2.jpeg"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_2.jpeg"
  />
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering jpg 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 1",
      "downloadUrl": "/url-path/fake_file_1.jpg",
      "name": "fake_file_1.jpg",
    }
  }
>
  <ImageRenderer
    fileName="fake_file_1.jpg"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_1.jpg"
  />
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering pdf 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 0",
      "downloadUrl": "/url-path/fake_file_0.pdf",
      "name": "fake_file_0.pdf",
    }
  }
>
  <PDFRenderer
    fileName="fake_file_0.pdf"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_0.pdf"
  />
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering png 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 4",
      "downloadUrl": "/url-path/fake_file_4.png",
      "name": "fake_file_4.png",
    }
  }
>
  <ImageRenderer
    fileName="fake_file_4.png"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_4.png"
  />
</FileCard>
`;

exports[`FileRenderer component snapshot successful rendering txt 1`] = `
<FileCard
  file={
    Object {
      "description": "file description 5",
      "downloadUrl": "/url-path/fake_file_5.txt",
      "name": "fake_file_5.txt",
    }
  }
>
  <TXTRenderer
    fileName="fake_file_5.txt"
    onError={[MockFunction onError]}
    onSuccess={[MockFunction onSuccess]}
    url="/url-path/fake_file_5.txt"
  />
</FileCard>
`;
